File: java/src/main/java/io/github/nickliapin/tdc/pattern/PatternGen.java
package io.github.nickliapin.tdc.pattern;

import io.github.nickliapin.tdc.prng.Prng;
import io.github.nickliapin.tdc.stats.Distribution;
import io.github.nickliapin.tdc.lib.Fixed;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * {@code <gen type="pattern" .../>} — data shaped like a drawing.
 *
 * <p>Three ways to read one:
 *
 * <ul>
 *   <li>{@code points="0,0 5,9 10,2"} — a single line, read as a trajectory. Deterministic: no
 *       draw is taken at all, so the column is the same under every seed.
 *   <li>{@code upper=} with an optional {@code lower=} — a band. Each row is a random value
 *       between the two lines, one draw apiece.
 *   <li>{@code mode="density"} — the same drawing read as a distribution instead.
 * </ul>
 *
 * <p>{@code spread="N"} widens a single line into a tunnel of ±N without drawing its edges by
 * hand, which is the usual way to turn a clean trend into something that looks measured.
 *
 * <p>Like the counters, a signal's value comes from the absolute row index rather than from the
 * row before it.
 */
public final class PatternGen {

  private static final Pattern NUMBER =
      Pattern.compile("-?\\d+(?:\\.\\d+)?(?:[eE][+-]?\\d+)?");

  private enum Kind {
    SIGNAL,
    CORRIDOR,
    DENSITY
  }

  private final Kind kind;
  private final Curve curve;
  private final Curve lower;
  private final Curve upper;
  private final Density density;
  private final double spread;
  private final int decimals;

  private PatternGen(
      Kind kind, Curve curve, Curve lower, Curve upper, Density density, double spread, int decimals) {
    this.kind = kind;
    this.curve = curve;
    this.lower = lower;
    this.upper = upper;
    this.density = density;
    this.spread = spread;
    this.decimals = decimals;
  }

  public static List<String> generate(
      Map<String, String> attrs, int count, java.nio.file.Path baseDir, Prng.Sfc32 prng) {
    PatternGen gen = of(attrs, baseDir);
    boolean draws = gen.draws();
    List<String> out = new ArrayList<>(count);
    // The drawing is stretched over the run: row i reads it at i/(count-1), and one row covers
    // 1/(count-1) of its width.
    double denom = count > 1 ? count - 1 : 1;
    for (int i = 0; i < count; i++) {
      double u = draws ? Distribution.openUnit(prng.next()) : 0;
      out.add(gen.valueAt(i / denom, u, 1 / denom));
    }
    return out;
  }

  static PatternGen of(Map<String, String> attrs) {
    return of(attrs, null);
  }

  public static PatternGen of(Map<String, String> attrs, java.nio.file.Path baseDir) {
    return of(attrs, baseDir, java.util.List.of());
  }

  public static PatternGen of(
      Map<String, String> attrs, java.nio.file.Path baseDir, java.util.List<java.nio.file.Path> roots) {
    double spread = spread(attrs);
    int decimals = decimals(attrs);
    double[] yRange = yRange(attrs.get("y_range"));
    Curve.Interp interp = interp(attrs.get("interp"));

    PatternGen gen;
    String upperRaw = attrs.get("upper");
    if (upperRaw != null && !upperRaw.isBlank()) {
      List<double[]> upperPts = points(upperRaw);
      String lowerRaw = attrs.get("lower");
      List<double[]> lowerPts =
          lowerRaw != null && !lowerRaw.isBlank() ? points(lowerRaw) : null;
      gen = corridor(upperPts, lowerPts, yRange, decimals, interp, spread);
    } else {
      String pointsRaw = attrs.get("points");
      if (pointsRaw == null || pointsRaw.isBlank()) {
        String src = attrs.get("src");
        if (src == null || src.isBlank()) {
          throw new IllegalArgumentException(
              "pattern: needs \"points\"/\"src\", or \"upper\"[/\"lower\"]");
        }
        gen = fromFile(src.trim(), baseDir, roots, attrs, yRange, decimals, interp, spread);
      } else {
        Curve c = Curve.of(points(pointsRaw), yRange, decimals, null, interp);
        gen = new PatternGen(Kind.SIGNAL, c, null, null, null, spread, decimals);
      }
    }

    if (!"density".equals(mode(attrs.get("mode")))) {
      return gen;
    }
    if (spread > 0) {
      throw new IllegalArgumentException(
          "pattern: \"spread\" has no meaning with mode=\"density\" — the drawing itself sets the scatter");
    }
    // A band contributes its top edge: the outline is the distribution, whatever its floor does.
    Curve source = gen.kind == Kind.CORRIDOR ? gen.upper : gen.curve;
    return new PatternGen(Kind.DENSITY, null, null, null, Density.of(source), 0, decimals);
  }

  /**
   * A drawing on disk: a picture, or a vector file.
   *
   * <p>Both are measured the same way — highest and lowest ink at each position — so one stroke
   * gives an exact curve and two strokes, or a closed outline, give a band. A file may switch
   * between the two along its own length, and a sketch usually does.
   */
  private static PatternGen fromFile(
      String src,
      java.nio.file.Path baseDir,
      java.util.List<java.nio.file.Path> roots,
      Map<String, String> attrs,
      double[] yRange,
      int decimals,
      Curve.Interp interp,
      double spread) {
    // The same resolution a file source gets, so a drawing may live in a data folder too.
    java.nio.file.Path path =
        io.github.nickliapin.tdc.generators.FileGen.resolve(src, baseDir, roots);
    byte[] bytes;
    try {
      bytes = java.nio.file.Files.readAllBytes(path);
    } catch (java.io.IOException e) {
      throw new java.io.UncheckedIOException("pattern: cannot read \"" + src + "\"", e);
    }

    if (Png.isPng(bytes)) {
      Png.Image image = Png.decode(bytes);
      SvgPath.Envelope traced = Png.trace(image, inkThreshold(attrs));
      // The frame is the value scale: the picture's own height is what 0..max means.
      return fromEnvelope(
          traced, yRange, decimals, interp, spread, new double[] {0, image.height() - 1});
    }

    SvgPath.Envelope envelope =
        SvgPath.envelope(new String(bytes, java.nio.charset.StandardCharsets.UTF_8), 600);
    return fromEnvelope(envelope, yRange, decimals, interp, spread, null);
  }

  /** A traced envelope becomes a plain line when its two edges coincide, and a band otherwise. */
  private static PatternGen fromEnvelope(
      SvgPath.Envelope envelope,
      double[] yRange,
      int decimals,
      Curve.Interp interp,
      double spread,
      double[] normExtent) {
    List<double[]> top = envelope.top();
    List<double[]> bottom = envelope.bottom();

    boolean banded = false;
    int n = Math.min(top.size(), bottom.size());
    for (int i = 0; i < n; i++) {
      if (Math.abs(top.get(i)[1] - bottom.get(i)[1]) > 1e-9) {
        banded = true;
        break;
      }
    }
    if (!banded) {
      return new PatternGen(
          Kind.SIGNAL, Curve.of(top, yRange, decimals, normExtent, interp), null, null, null,
          spread, decimals);
    }

    double lo = Double.POSITIVE_INFINITY;
    double hi = Double.NEGATIVE_INFINITY;
    for (double[] p : top) {
      lo = Math.min(lo, p[1]);
      hi = Math.max(hi, p[1]);
    }
    for (double[] p : bottom) {
      lo = Math.min(lo, p[1]);
      hi = Math.max(hi, p[1]);
    }
    double[] extent = normExtent != null ? normExtent : new double[] {lo, hi};
    return new PatternGen(
        Kind.CORRIDOR, null,
        Curve.of(bottom, yRange, decimals, extent, interp),
        Curve.of(top, yRange, decimals, extent, interp),
        null, spread, decimals);
  }

  /** {@code ink_threshold} — how dark counts as a line, on an opaque background. */
  private static double inkThreshold(Map<String, String> attrs) {
    String raw = attrs.get("ink_threshold");
    if (raw == null || raw.isBlank()) {
      return 0.5;
    }
    try {
      double t = Double.parseDouble(raw.trim());
      if (!Double.isFinite(t) || t <= 0 || t >= 1) {
        throw new NumberFormatException();
      }
      return t;
    } catch (NumberFormatException e) {
      throw new IllegalArgumentException(
          "pattern: \"ink_threshold\" must be a number strictly between 0 and 1");
    }
  }

  /**
   * A corridor: two lines in one value space.
   *
   * <p>Both are normalized against their <em>shared</em> height extent, so the band between them
   * means something. Normalizing each against its own extent would stretch them to the same
   * height and collapse the corridor.
   */
  private static PatternGen corridor(
      List<double[]> upperPts,
      List<double[]> lowerPts,
      double[] yRange,
      int decimals,
      Curve.Interp interp,
      double spread) {
    List<Double> heights = new ArrayList<>();
    for (double[] p : upperPts) {
      heights.add(p[1]);
    }
    if (lowerPts != null) {
      for (double[] p : lowerPts) {
        heights.add(p[1]);
      }
    } else {
      // No lower line means a flat floor at zero, which belongs in the shared extent.
      heights.add(0.0);
    }
    double lo = heights.stream().mapToDouble(Double::doubleValue).min().orElse(0);
    double hi = heights.stream().mapToDouble(Double::doubleValue).max().orElse(0);
    double[] extent = {lo, hi};

    Curve upper = Curve.of(upperPts, yRange, decimals, extent, interp);
    Curve lower;
    if (lowerPts != null) {
      lower = Curve.of(lowerPts, yRange, decimals, extent, interp);
    } else {
      double x0 = upperPts.get(0)[0];
      double xN = upperPts.get(upperPts.size() - 1)[0];
      lower = Curve.of(List.of(new double[] {x0, lo}, new double[] {xN, lo}), yRange, decimals, extent, interp);
    }
    return new PatternGen(Kind.CORRIDOR, null, lower, upper, null, spread, decimals);
  }

  /** Whether a row costs a draw: a band, a spread, or a density. A bare line costs none. */
  public boolean draws() {
    return kind != Kind.SIGNAL || spread > 0;
  }

  public String valueAt(double t, double u, double dt) {
    if (kind == Kind.DENSITY) {
      // Position in the run means nothing here — the drawing is a distribution, so the row's
      // own draw picks the value and the order comes out random.
      return format(density.valueAt(u), density.decimals());
    }
    if (kind == Kind.SIGNAL) {
      double v = curve.valueAt(t, dt);
      return format(spread > 0 ? v + (2 * u - 1) * spread : v, decimals);
    }
    double a = lower.valueAt(t, dt);
    double b = upper.valueAt(t, dt);
    double lo = Math.min(a, b) - spread;
    double hi = Math.max(a, b) + spread;
    return format(lo + u * (hi - lo), decimals);
  }

  // ── attributes ───────────────────────────────────────────────────────────────────────────

  /** Every number in the text, in pairs. Whatever separates them is decoration. */
  static List<double[]> points(String raw) {
    Matcher m = NUMBER.matcher(raw);
    List<Double> nums = new ArrayList<>();
    while (m.find()) {
      nums.add(Double.parseDouble(m.group()));
    }
    if (nums.isEmpty() || nums.size() % 2 != 0) {
      throw new IllegalArgumentException(
          "pattern: points must be an even list of \"x,y\" coordinates (got "
              + nums.size()
              + " numbers)");
    }
    List<double[]> points = new ArrayList<>(nums.size() / 2);
    for (int i = 0; i < nums.size(); i += 2) {
      points.add(new double[] {nums.get(i), nums.get(i + 1)});
    }
    return points;
  }

  static double[] yRange(String raw) {
    if (raw == null || raw.isBlank()) {
      return null;
    }
    String[] parts = raw.split("\\.\\.", -1);
    if (parts.length != 2) {
      throw new IllegalArgumentException(
          "pattern: y_range \"" + raw + "\" must be \"min..max\" with two numbers");
    }
    try {
      double a = Double.parseDouble(parts[0].trim());
      double b = Double.parseDouble(parts[1].trim());
      if (!Double.isFinite(a) || !Double.isFinite(b)) {
        throw new NumberFormatException();
      }
      return new double[] {a, b};
    } catch (NumberFormatException e) {
      throw new IllegalArgumentException(
          "pattern: y_range \"" + raw + "\" must be \"min..max\" with two numbers");
    }
  }

  public static Curve.Interp interp(String raw) {
    if (raw == null || raw.isBlank()) {
      return Curve.Interp.LINEAR;
    }
    return switch (raw.trim().toLowerCase()) {
      case "linear" -> Curve.Interp.LINEAR;
      case "smooth" -> Curve.Interp.SMOOTH;
      case "step" -> Curve.Interp.STEP;
      default -> throw new IllegalArgumentException(
          "pattern: \"interp\" must be \"linear\", \"smooth\" or \"step\"");
    };
  }

  public static String mode(String raw) {
    if (raw == null || raw.isBlank()) {
      return "signal";
    }
    String v = raw.trim().toLowerCase();
    if (!"signal".equals(v) && !"density".equals(v)) {
      throw new IllegalArgumentException(
          "pattern: \"mode\" must be \"signal\" (a trajectory) or \"density\" (a distribution)");
    }
    return v;
  }

  public static double spread(Map<String, String> attrs) {
    String raw = attrs.get("spread");
    if (raw == null || raw.isBlank()) {
      return 0;
    }
    try {
      double s = Double.parseDouble(raw.trim());
      if (!Double.isFinite(s) || s < 0) {
        throw new NumberFormatException();
      }
      return s;
    } catch (NumberFormatException e) {
      throw new IllegalArgumentException("pattern: \"spread\" must be a non-negative number");
    }
  }

  public static int decimals(Map<String, String> attrs) {
    String raw = attrs.get("decimals");
    if (raw == null || raw.isBlank()) {
      return 0;
    }
    try {
      int d = Integer.parseInt(raw.trim());
      if (d < 0) {
        throw new NumberFormatException();
      }
      return d;
    } catch (NumberFormatException e) {
      throw new IllegalArgumentException("pattern: \"decimals\" must be a non-negative integer");
    }
  }

  private static String format(double v, int decimals) {
    return Fixed.toFixed(v, decimals);
  }
}

File: csharp/Tdcv2/Pattern/PatternGen.cs
using System.Globalization;
using System.Text.RegularExpressions;
using Tdcv2.Prng;

namespace Tdcv2.Pattern;

/// <summary>
/// <c>&lt;gen type="pattern" .../&gt;</c> — data shaped like a drawing.
/// </summary>
/// <remarks>
/// <para>Three ways to read one:</para>
/// <list type="bullet">
///   <item>
///     <c>points="0,0 5,9 10,2"</c> — a single line, read as a trajectory. Deterministic: no draw
///     is taken at all, so the column is the same under every seed.
///   </item>
///   <item>
///     <c>upper=</c> with an optional <c>lower=</c> — a band. Each row is a random value between
///     the two lines, one draw apiece.
///   </item>
///   <item><c>mode="density"</c> — the same drawing read as a distribution instead.</item>
/// </list>
/// <para>
/// <c>spread="N"</c> widens a single line into a tunnel of ±N without drawing its edges by hand,
/// which is the usual way to turn a clean trend into something that looks measured.
/// </para>
/// <para>
/// Like the counters, a signal's value comes from the absolute row index rather than from the row
/// before it.
/// </para>
/// <para>
/// A drawing loaded from a file — <c>src=</c> pointing at a PNG or an SVG — is read the same way
/// either format: highest and lowest ink at each position, so one stroke is an exact curve and two
/// strokes are a band.
/// </para>
/// </remarks>
public sealed class PatternGen
{
    private static readonly Regex Number =
        new(@"-?\d+(?:\.\d+)?(?:[eE][+-]?\d+)?", RegexOptions.Compiled);

    private enum Kind
    {
        Signal,
        Corridor,
        DensityKind,
    }

    private readonly Kind _kind;
    private readonly Curve? _curve;
    private readonly Curve? _lower;
    private readonly Curve? _upper;
    private readonly Density? _density;
    private readonly double _spread;
    private readonly int _decimals;

    private PatternGen(
        Kind kind, Curve? curve, Curve? lower, Curve? upper, Density? density, double spread,
        int decimals)
    {
        _kind = kind;
        _curve = curve;
        _lower = lower;
        _upper = upper;
        _density = density;
        _spread = spread;
        _decimals = decimals;
    }

    public static IReadOnlyList<string> Generate(
        IReadOnlyDictionary<string, string> attrs, int count, Sfc32 prng, string? baseDir = null,
        IReadOnlyList<string>? roots = null)
    {
        PatternGen gen = Of(attrs, baseDir, roots);
        bool draws = gen.Draws;
        var result = new List<string>(count);
        // The drawing is stretched over the run: row i reads it at i/(count-1), and one row covers
        // 1/(count-1) of its width.
        double denom = count > 1 ? count - 1 : 1;
        for (int i = 0; i < count; i++)
        {
            double u = draws ? Seekable.OpenUnit(prng.Next()) : 0;
            result.Add(gen.ValueAt(i / denom, u));
        }

        return result;
    }

    public static PatternGen Of(
        IReadOnlyDictionary<string, string> attrs, string? baseDir = null,
        IReadOnlyList<string>? roots = null)
    {
        double spread = Spread(attrs);
        int decimals = DecimalsOf(attrs);
        double[]? yRange = YRange(attrs.GetValueOrDefault("y_range"));
        Interp interp = InterpOf(attrs.GetValueOrDefault("interp"));

        PatternGen gen;
        string? upperRaw = attrs.GetValueOrDefault("upper");
        if (!string.IsNullOrWhiteSpace(upperRaw))
        {
            IReadOnlyList<double[]> upperPts = Points(upperRaw);
            string? lowerRaw = attrs.GetValueOrDefault("lower");
            IReadOnlyList<double[]>? lowerPts =
                string.IsNullOrWhiteSpace(lowerRaw) ? null : Points(lowerRaw);
            gen = Corridor(upperPts, lowerPts, yRange, decimals, interp, spread);
        }
        else
        {
            string? pointsRaw = attrs.GetValueOrDefault("points");
            if (string.IsNullOrWhiteSpace(pointsRaw))
            {
                string? src = attrs.GetValueOrDefault("src");
                if (string.IsNullOrWhiteSpace(src))
                {
                    throw new ArgumentException(
                        "pattern: needs \"points\"/\"src\", or \"upper\"[/\"lower\"]");
                }

                gen = FromFile(src.Trim(), baseDir, roots, attrs, yRange, decimals, interp, spread);
            }
            else
            {
                Curve c = Curve.Of(Points(pointsRaw), yRange, decimals, null, interp);
                gen = new PatternGen(Kind.Signal, c, null, null, null, spread, decimals);
            }
        }

        if (Mode(attrs.GetValueOrDefault("mode")) != "density")
        {
            return gen;
        }

        if (spread > 0)
        {
            throw new ArgumentException(
                "pattern: \"spread\" has no meaning with mode=\"density\" — the drawing itself "
                + "sets the scatter");
        }

        // A band contributes its top edge: the outline is the distribution, whatever its floor does.
        Curve source = gen._kind == Kind.Corridor ? gen._upper! : gen._curve!;
        return new PatternGen(
            Kind.DensityKind, null, null, null, Density.Of(source), 0, decimals);
    }

    /// <summary>
    /// A drawing on disk: a picture, or a vector file.
    /// </summary>
    /// <remarks>
    /// Both are measured the same way — highest and lowest ink at each position — so one stroke
    /// gives an exact curve and two strokes, or a closed outline, give a band. A file may switch
    /// between the two along its own length, and a sketch usually does.
    /// </remarks>
    private static PatternGen FromFile(
        string src, string? baseDir, IReadOnlyList<string>? roots,
        IReadOnlyDictionary<string, string> attrs, double[]? yRange, int decimals, Interp interp,
        double spread)
    {
        // The same resolution a file source gets, so a drawing may live in a data folder too.
        string path = Generators.FileGen.Resolve(src, baseDir, roots);
        byte[] bytes;
        try
        {
            bytes = File.ReadAllBytes(path);
        }
        catch (IOException e)
        {
            throw new ArgumentException($"pattern: cannot read \"{src}\"", e);
        }
        catch (UnauthorizedAccessException e)
        {
            throw new ArgumentException($"pattern: cannot read \"{src}\"", e);
        }

        if (Png.IsPng(bytes))
        {
            Png.Image image = Png.Decode(bytes);
            SvgPath.Envelope traced = Png.Trace(image, InkThreshold(attrs));
            // The frame is the value scale: the picture's own height is what 0..max means.
            return FromEnvelope(
                traced, yRange, decimals, interp, spread, new double[] { 0, image.Height - 1 });
        }

        SvgPath.Envelope envelope =
            SvgPath.GetEnvelope(System.Text.Encoding.UTF8.GetString(bytes), 600);
        return FromEnvelope(envelope, yRange, decimals, interp, spread, null);
    }

    /// <summary>
    /// A traced envelope becomes a plain line when its two edges coincide, and a band otherwise.
    /// </summary>
    private static PatternGen FromEnvelope(
        SvgPath.Envelope envelope, double[]? yRange, int decimals, Interp interp, double spread,
        double[]? normExtent)
    {
        IReadOnlyList<double[]> top = envelope.Top;
        IReadOnlyList<double[]> bottom = envelope.Bottom;

        bool banded = false;
        int n = Math.Min(top.Count, bottom.Count);
        for (int i = 0; i < n; i++)
        {
            if (Math.Abs(top[i][1] - bottom[i][1]) > 1e-9)
            {
                banded = true;
                break;
            }
        }

        if (!banded)
        {
            return new PatternGen(
                Kind.Signal, Curve.Of(top, yRange, decimals, normExtent, interp), null, null, null,
                spread, decimals);
        }

        double lo = double.PositiveInfinity;
        double hi = double.NegativeInfinity;
        foreach (double[] p in top)
        {
            lo = Math.Min(lo, p[1]);
            hi = Math.Max(hi, p[1]);
        }

        foreach (double[] p in bottom)
        {
            lo = Math.Min(lo, p[1]);
            hi = Math.Max(hi, p[1]);
        }

        double[] extent = normExtent ?? new[] { lo, hi };
        return new PatternGen(
            Kind.Corridor,
            null,
            Curve.Of(bottom, yRange, decimals, extent, interp),
            Curve.Of(top, yRange, decimals, extent, interp),
            null,
            spread,
            decimals);
    }

    /// <summary><c>ink_threshold</c> — how dark counts as a line, on an opaque background.</summary>
    private static double InkThreshold(IReadOnlyDictionary<string, string> attrs)
    {
        string? raw = attrs.GetValueOrDefault("ink_threshold");
        if (string.IsNullOrWhiteSpace(raw))
        {
            return 0.5;
        }

        if (!double.TryParse(
                raw.Trim(), NumberStyles.Float, CultureInfo.InvariantCulture, out double t)
            || !double.IsFinite(t) || t <= 0 || t >= 1)
        {
            throw new ArgumentException(
                "pattern: \"ink_threshold\" must be a number strictly between 0 and 1");
        }

        return t;
    }

    /// <summary>
    /// A corridor: two lines in one value space.
    /// </summary>
    /// <remarks>
    /// Both are normalized against their <em>shared</em> height extent, so the band between them
    /// means something. Normalizing each against its own extent would stretch them to the same
    /// height and collapse the corridor.
    /// </remarks>
    private static PatternGen Corridor(
        IReadOnlyList<double[]> upperPts, IReadOnlyList<double[]>? lowerPts, double[]? yRange,
        int decimals, Interp interp, double spread)
    {
        var heights = new List<double>();
        foreach (double[] p in upperPts)
        {
            heights.Add(p[1]);
        }

        if (lowerPts is not null)
        {
            foreach (double[] p in lowerPts)
            {
                heights.Add(p[1]);
            }
        }
        else
        {
            // No lower line means a flat floor at zero, which belongs in the shared extent.
            heights.Add(0.0);
        }

        double lo = heights.Count == 0 ? 0 : heights.Min();
        double hi = heights.Count == 0 ? 0 : heights.Max();
        var extent = new[] { lo, hi };

        Curve upper = Curve.Of(upperPts, yRange, decimals, extent, interp);
        Curve lower;
        if (lowerPts is not null)
        {
            lower = Curve.Of(lowerPts, yRange, decimals, extent, interp);
        }
        else
        {
            double x0 = upperPts[0][0];
            double xN = upperPts[^1][0];
            lower = Curve.Of(
                new[] { new[] { x0, lo }, new[] { xN, lo } }, yRange, decimals, extent, interp);
        }

        return new PatternGen(Kind.Corridor, null, lower, upper, null, spread, decimals);
    }

    /// <summary>Whether a row costs a draw: a band, a spread, or a density. A bare line costs none.</summary>
    public bool Draws => _kind != Kind.Signal || _spread > 0;

    public string ValueAt(double t, double u)
    {
        if (_kind == Kind.DensityKind)
        {
            // Position in the run means nothing here — the drawing is a distribution, so the row's
            // own draw picks the value and the order comes out random.
            return Format(_density!.ValueAt(u), _density.Decimals);
        }

        if (_kind == Kind.Signal)
        {
            double v = _curve!.ValueAt(t);
            return Format(_spread > 0 ? v + (((2 * u) - 1) * _spread) : v, _decimals);
        }

        double a = _lower!.ValueAt(t);
        double b = _upper!.ValueAt(t);
        double low = Math.Min(a, b) - _spread;
        double high = Math.Max(a, b) + _spread;
        return Format(low + (u * (high - low)), _decimals);
    }

    // ── attributes ───────────────────────────────────────────────────────────────────────────

    /// <summary>Every number in the text, in pairs. Whatever separates them is decoration.</summary>
    internal static IReadOnlyList<double[]> Points(string raw)
    {
        // A `;` is NOT a separator: every number is read in order, so
        // `0,20 100,20; 0,80 100,80` would silently become ONE curve of four points.
        // Somebody writing that meant a band, so name the spelling that works.
        if (raw.Contains(';', StringComparison.Ordinal))
        {
            throw new ArgumentException(
                "pattern: \";\" does not separate two lines in points= — every number is read "
                + "as one curve. For a band, draw the two edges separately: "
                + "upper=\"0,80 100,80\" lower=\"0,20 100,20\".");
        }

        var nums = new List<double>();
        foreach (Match m in Number.Matches(raw))
        {
            nums.Add(double.Parse(m.Value, CultureInfo.InvariantCulture));
        }

        if (nums.Count == 0 || nums.Count % 2 != 0)
        {
            throw new ArgumentException(
                "pattern: points must be an even list of \"x,y\" coordinates (got "
                + $"{nums.Count} numbers)");
        }

        var points = new List<double[]>(nums.Count / 2);
        for (int i = 0; i < nums.Count; i += 2)
        {
            points.Add(new[] { nums[i], nums[i + 1] });
        }

        return points;
    }

    /// <summary>
    /// <c>y_range="min..max"</c> — the value axis, and REQUIRED. A drawing carries no units of
    /// its own: a curve exported from one tool runs 0..100, from another 0..10002345345. Without
    /// it every answer would be a guess about somebody's export settings.
    /// </summary>
    internal static double[]? YRange(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            throw new ArgumentException(
                "pattern: y_range is required — it is the value axis a drawing is brought into, "
                + "and a drawing has no scale of its own. Write y_range=\"0..100\".");
        }

        string[] parts = raw.Split("..");
        if (parts.Length != 2
            || !double.TryParse(parts[0].Trim(), NumberStyles.Float, CultureInfo.InvariantCulture, out double a)
            || !double.TryParse(parts[1].Trim(), NumberStyles.Float, CultureInfo.InvariantCulture, out double b)
            || !double.IsFinite(a) || !double.IsFinite(b))
        {
            throw new ArgumentException(
                $"pattern: y_range \"{raw}\" must be \"min..max\" with two numbers");
        }

        return new[] { a, b };
    }

    internal static Interp InterpOf(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            return Interp.Linear;
        }

        return raw.Trim().ToLowerInvariant() switch
        {
            "linear" => Interp.Linear,
            "smooth" => Interp.Smooth,
            "step" => Interp.Step,
            _ => throw new ArgumentException(
                "pattern: \"interp\" must be \"linear\", \"smooth\" or \"step\""),
        };
    }

    internal static string Mode(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            return "signal";
        }

        string v = raw.Trim().ToLowerInvariant();
        if (v != "signal" && v != "density")
        {
            throw new ArgumentException(
                "pattern: \"mode\" must be \"signal\" (a trajectory) or \"density\" (a distribution)");
        }

        return v;
    }

    internal static double Spread(IReadOnlyDictionary<string, string> attrs)
    {
        string? raw = attrs.GetValueOrDefault("spread");
        if (string.IsNullOrWhiteSpace(raw))
        {
            return 0;
        }

        if (!double.TryParse(raw.Trim(), NumberStyles.Float, CultureInfo.InvariantCulture, out double s)
            || !double.IsFinite(s) || s < 0)
        {
            throw new ArgumentException("pattern: \"spread\" must be a non-negative number");
        }

        return s;
    }

    internal static int DecimalsOf(IReadOnlyDictionary<string, string> attrs)
    {
        string? raw = attrs.GetValueOrDefault("decimals");
        if (string.IsNullOrWhiteSpace(raw))
        {
            return 0;
        }

        if (!int.TryParse(raw.Trim(), NumberStyles.Integer, CultureInfo.InvariantCulture, out int d)
            || d < 0)
        {
            throw new ArgumentException("pattern: \"decimals\" must be a non-negative integer");
        }

        return d;
    }

    private static string Format(double v, int decimals) =>
        Stats.Distribution.ToFixed(v, decimals);
}
